src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabWebHookListener.java

```diff
@@ -39,27 +39,49 @@ public void onNoteEvent(NoteEvent noteEvent) {
     @Override
     public void onMergeRequestEvent(MergeRequestEvent mrEvent) {
         LOGGER.log(Level.FINE, mrEvent.toString());
-        final long triggerDelay = findTriggerDelay(mrEvent.getProject().getWebUrl());
         GitLabMergeRequestSCMEvent trigger = new GitLabMergeRequestSCMEvent(mrEvent, origin);
+        if (findImmediateHookTrigger(mrEvent.getProject().getWebUrl())) {
+            SCMHeadEvent.fireNow(trigger);
+        }
+        final long triggerDelay = findTriggerDelay(mrEvent.getProject().getWebUrl());
         SCMHeadEvent.fireLater(trigger, triggerDelay, TimeUnit.SECONDS);
     }
 
     @Override
     public void onPushEvent(PushEvent pushEvent) {
         LOGGER.log(Level.FINE, pushEvent.toString());
-        final long triggerDelay = findTriggerDelay(pushEvent.getProject().getWebUrl());
         GitLabPushSCMEvent trigger = new GitLabPushSCMEvent(pushEvent, origin);
+        if (findImmediateHookTrigger(pushEvent.getProject().getWebUrl())) {
+            SCMHeadEvent.fireNow(trigger);
+        }
+        final long triggerDelay = findTriggerDelay(pushEvent.getProject().getWebUrl());
         SCMHeadEvent.fireLater(trigger, triggerDelay, TimeUnit.SECONDS);
     }
 
     @Override
     public void onTagPushEvent(TagPushEvent tagPushEvent) {
         LOGGER.log(Level.FINE, tagPushEvent.toString());
-        final long triggerDelay = findTriggerDelay(tagPushEvent.getProject().getWebUrl());
         GitLabTagPushSCMEvent trigger = new GitLabTagPushSCMEvent(tagPushEvent, origin);
+        if (findImmediateHookTrigger(tagPushEvent.getProject().getWebUrl())) {
+            SCMHeadEvent.fireNow(trigger);
+        }
+        final long triggerDelay = findTriggerDelay(tagPushEvent.getProject().getWebUrl());
         SCMHeadEvent.fireLater(trigger, triggerDelay, TimeUnit.SECONDS);
     }
 
+    private boolean findImmediateHookTrigger(final String projectUrl) {
+        final GitLabServer projectServer = findProjectServer(projectUrl);
```

Comment:
I think we need to reduce the amount of calls for this findProjectServer.

Maybe pass the GitLabServer object instead.

Reply:
I've refactored the class to do a single call to findProjectServer as suggested (per trigger) and also reduce the code duplicates between the different trigger types.